src/Features/LanguageServer/Protocol/Handler/Diagnostics/DocumentPullDiagnosticHandler.cs

```diff
@@ -39,10 +53,18 @@ protected override ImmutableArray<Document> GetOrderedDocuments(RequestContext c
             return context.Document == null ? ImmutableArray<Document>.Empty : ImmutableArray.Create(context.Document);
         }
 
-        protected override DiagnosticParams[]? GetPreviousResults(DocumentDiagnosticsParams diagnosticsParams)
-            => new[] { diagnosticsParams };
+        protected override Task<ImmutableArray<DiagnosticData>> GetDiagnosticsAsync(
+            Document document, Option2<DiagnosticMode> diagnosticMode, CancellationToken cancellationToken)
+        {
+            // We only support doc diagnostics for open files.
+            if (!document.IsOpen())
```

Comment:
Is this going to work for Razor? Does the LSP didOpen handler need to do something to tell the workspace the document is open (as part of the server per workspace changes, not this PR)?

Reply:
i've switched this to be based on LSP semantics now.